File: 0046-Permutations/lc0046.rs
// Rust solution.

impl Solution {
    pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut ans: Vec<Vec<i32>> = Vec::new();
        let mut res: Vec<i32> = Vec::new();
        let mut used = vec![false; nums.len()];
        Self::backtrack(&nums, &mut res, &mut used, &mut ans);
        ans
    }

    pub fn backtrack(nums: &Vec<i32>, res: &mut Vec<i32>,
                     used: &mut Vec<bool>, ans: &mut Vec<Vec<i32>>) -> () {
        if res.len() == nums.len() {
            ans.push(res.clone());
        } else {
            for i in 0..nums.len() {
                if !used[i] {
                    res.push(nums[i]);
                    used[i] = true;
                    Self::backtrack(nums, res, used, ans);
                    res.pop();
                    used[i] = false;
                }
            }
        }
    }
}

File: 0046-Permutations/lc0046.cc
// C++ solution.

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> res;
        vector<bool> used(nums.size(), false);
        backtrack(nums, res, used, ans);
        return ans;
    }

    void backtrack(vector<int>& nums, vector<int>& res, vector<bool>& used, vector<vector<int>>& ans) {
        if (res.size() == nums.size()) {
            ans.push_back(res);
        } else {
            for (int i = 0; i < nums.size(); i++) {
                if (!used[i]) {
                    res.push_back(nums[i]);
                    used[i] = true;
                    backtrack(nums, res, used, ans);
                    res.pop_back();
                    used[i] = false;
                }
            }
        }
    }
};
